src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs

```diff
@@ -300,15 +268,26 @@ internal override void DocumentOpened(string projectFilePath, string documentFil
                     });
                 }
 
-                _openDocuments.Add(documentFilePath);
+                ProjectChangeEventArgs eventArgs = null;
 
-                // Document updates can no-op.
-                if (!object.ReferenceEquals(state, entry.State))
+                using (var _ = upgradeableLock.GetWriteLock())
                 {
-                    var oldSnapshot = entry.GetSnapshot();
-                    entry = new Entry(state);
-                    _projects[projectFilePath] = entry;
-                    NotifyListeners(oldSnapshot, entry.GetSnapshot(), documentFilePath, ProjectChangeKind.DocumentChanged);
+                    _openDocuments_needsLock.Add(documentFilePath);
+
+                    // Document updates can no-op.
+                    if (!object.ReferenceEquals(state, entry.State))
+                    {
+                        var oldSnapshot = entry.GetSnapshot();
+                        entry = new Entry(state);
+                        _projects_needsLock[projectFilePath] = entry;
+                        eventArgs = new(oldSnapshot, entry.GetSnapshot(), documentFilePath, ProjectChangeKind.DocumentChanged, IsSolutionClosing);
+                    }
+                }
+
+                // Don't notify with a write lock to avoid reentrancy problems
```

Comment:
Seems like notifying under just a read lock could be an issue too, unless we're going to pass the lock through the EventArgs, otherwise what happens if a listener wants to upgrade again and make a change? Definitely feels like quite a lot of work could be getting done under the read, potentially.

Reply:
Potentially. I think auditing consumers of the notification will help here, and then maybe I can remove the lock for notification.